error: lockfile had changes, but lockfile is frozen

### What version of Bun is running?

1.0.2

### What platform is your computer?

Darwin 22.6.0 arm64 arm

### What steps can reproduce the bug?

I am using below command inside a docker image and it is failing with error - error: lockfile had changes, but lockfile is frozen
`bun install --production`

- Initially everything was working as expected
- I updated one of the packages
- and after that tried to build the docker image again.

I am using bun as a package manager.

node version - 18.16.0the 

### What is the expected behavior?

should install packages without any issues.

### What do you see instead?

```
 => ERROR [stage-1 6/8] RUN bun install --production                                                                                                                                                          0.2s 
------                                                                                                                                                                                                             
 > [stage-1 6/8] RUN bun install --production:                                                                                                                                                                     
0.209 bun install v1.0.2 (37edd5a6)                                                                                                                                                                                
0.211 error: lockfile had changes, but lockfile is frozen                                                                                                                                                          
------                                                                                                                                                                                                             
Dockerfile:24
--------------------
  22 |     COPY package.json bun.lockb .env.dev firebase-dev-admin.json ./
  23 |     RUN npm install -g bun
  24 | >>> RUN bun install --production
  25 |     
  26 |     # cleanup 
--------------------
ERROR: failed to solve: process "/bin/sh -c bun install --production" did not complete successfully: exit code: 1
```

### Additional information

I tried using `frozen-lockfile` flag as well, with no help

```
 => ERROR [stage-1 6/8] RUN bun install --production --frozen-lockfile                                                                                                                                        0.3s 
------                                                                                                                                                                                                             
 > [stage-1 6/8] RUN bun install --production --frozen-lockfile:                                                                                                                                                   
0.241 bun install v1.0.2 (37edd5a6)                                                                                                                                                                                
0.243 error: lockfile had changes, but lockfile is frozen                                                                                                                                                          
------                                                                                                                                                                                                             
Dockerfile:24
--------------------
  22 |     COPY package.json bun.lockb .env.dev firebase-dev-admin.json ./
  23 |     RUN npm install -g bun
  24 | >>> RUN bun install --production --frozen-lockfile
  25 |     
  26 |     # cleanup 
--------------------
ERROR: failed to solve: process "/bin/sh -c bun install --production --frozen-lockfile" did not complete successfully: exit code: 1
```

I'm getting this error in a container inside a completely empty directory when using `-p` (for production):
```bash
$ mkdir /tmp/bun-tmp
$ cd "$_"
$ bun install -p <some-dependency>
bun add v1.0.18 (36c316a2)
error: lockfile had changes, but lockfile is frozen
Error: Process completed with exit code 1.
```
Might be related to #1590.

--- a/src/install/install.zig
+++ b/src/install/install.zig
@@ -8410,12 +8410,14 @@ pub const PackageManager = struct {
         ctx: Command.Context,
         comptime log_level: PackageManager.Options.LogLevel,
     ) !PackageInstall.Summary {
-        var lockfile = this.lockfile;
+        const original_lockfile = this.lockfile;
+        defer this.lockfile = original_lockfile;
         if (!this.options.local_package_features.dev_dependencies) {
-            lockfile = try lockfile.maybeCloneFilteringRootPackages(
+            this.lockfile = try this.lockfile.maybeCloneFilteringRootPackages(
                 this.options.local_package_features,
                 this.options.enable.exact_versions,
             );
+            this.lockfile = this.lockfile;
         }
 
         const root_lifecycle_scripts_count = brk: {
@@ -8447,7 +8449,7 @@ pub const PackageManager = struct {
             progress.supports_ansi_escape_codes = Output.enable_ansi_colors_stderr;
             download_node = root_node.start(ProgressStrings.download(), 0);
 
-            install_node = root_node.start(ProgressStrings.install(), lockfile.packages.len);
+            install_node = root_node.start(ProgressStrings.install(), this.lockfile.packages.len);
             scripts_node = root_node.start(ProgressStrings.script(), root_lifecycle_scripts_count);
             this.downloads_node = &download_node;
             this.scripts_node = &scripts_node;
@@ -8499,7 +8501,7 @@ pub const PackageManager = struct {
         var summary = PackageInstall.Summary{};
 
         {
-            var iterator = Lockfile.Tree.Iterator.init(lockfile);
+            var iterator = Lockfile.Tree.Iterator.init(this.lockfile);
             if (comptime Environment.isPosix) {
                 Bin.Linker.ensureUmask();
             }
@@ -8507,10 +8509,10 @@ pub const PackageManager = struct {
                 // These slices potentially get resized during iteration
                 // so we want to make sure they're not accessible to the rest of this function
                 // to make mistakes harder
-                var parts = lockfile.packages.slice();
+                var parts = this.lockfile.packages.slice();
 
                 const completed_trees, const tree_ids_to_trees_the_id_depends_on, const tree_install_counts = trees: {
-                    const trees = lockfile.buffers.trees.items;
+                    const trees = this.lockfile.buffers.trees.items;
                     const completed_trees = try Bitset.initEmpty(this.allocator, trees.len);
                     var tree_ids_to_trees_the_id_depends_on = try Bitset.List.initEmpty(this.allocator, trees.len, trees.len);
 
@@ -8565,7 +8567,7 @@ pub const PackageManager = struct {
                     .root_node_modules_folder = node_modules_folder,
                     .names = parts.items(.name),
                     .resolutions = parts.items(.resolution),
-                    .lockfile = lockfile,
+                    .lockfile = this.lockfile,
                     .node = &install_node,
                     .node_modules_folder = node_modules_folder,
                     .progress = progress,
@@ -8576,7 +8578,7 @@ pub const PackageManager = struct {
                     .force_install = options.enable.force_install,
                     .successfully_installed = try Bitset.initEmpty(
                         this.allocator,
-                        lockfile.packages.len,
+                        this.lockfile.packages.len,
                     ),
                     .tree_iterator = &iterator,
                     .command_ctx = ctx,
@@ -8697,7 +8699,7 @@ pub const PackageManager = struct {
             }
 
             if (comptime Environment.allow_assert) {
-                for (lockfile.buffers.trees.items) |tree| {
+                for (this.lockfile.buffers.trees.items) |tree| {
                     std.debug.assert(installer.tree_install_counts[tree.id] == tree.dependencies.len);
                 }
             }
@@ -9133,7 +9135,7 @@ pub const PackageManager = struct {
             manager.lockfile.verifyResolutions(manager.options.local_package_features, manager.options.remote_package_features, log_level);
         }
 
-        if (needs_clean_lockfile or manager.options.enable.force_save_lockfile) {
+        if (needs_clean_lockfile or manager.options.enable.force_save_lockfile or manager.options.enable.frozen_lockfile) {
             did_meta_hash_change = try manager.lockfile.hasMetaHashChanged(
                 PackageManager.verbose_install or manager.options.do.print_meta_hash_string,
             );

--- a/src/install/install.zig
+++ b/src/install/install.zig
@@ -8871,13 +8871,6 @@ pub const PackageManager = struct {
 
                     had_any_diffs = had_any_diffs or manager.summary.hasDiffs();
 
-                    if (manager.options.enable.frozen_lockfile and had_any_diffs) {
-                        if (comptime log_level != .silent) {
-                            Output.prettyErrorln("<r><red>error<r>: lockfile had changes, but lockfile is frozen", .{});
-                        }
-                        Global.crash();
-                    }
-
                     if (manager.summary.new_trusted_dependencies.count() > 0) {
                         needs_new_lockfile = true;
                     }
@@ -9148,6 +9141,14 @@ pub const PackageManager = struct {
 
         if (manager.options.global) {
             try manager.setupGlobalDir(&ctx);
+        }
+
+        if (did_meta_hash_change and manager.options.enable.frozen_lockfile) {
+            if (comptime log_level != .silent) {
+                Output.prettyErrorln("<r><red>error<r><d>:<r> lockfile had changes, but lockfile is frozen", .{});
+                Output.note("try re-running without <d>--frozen-lockfile<r> and commit the updated lockfile", .{});
+            }
+            Global.crash();
         }
 
         // It's unnecessary work to re-save the lockfile if there are no changes

--- a/src/install/install.zig
+++ b/src/install/install.zig
@@ -8417,7 +8417,6 @@ pub const PackageManager = struct {
                 this.options.local_package_features,
                 this.options.enable.exact_versions,
             );
-            this.lockfile = this.lockfile;
         }
 
         const root_lifecycle_scripts_count = brk: {

--- a/src/install/lockfile.zig
+++ b/src/install/lockfile.zig
@@ -3087,6 +3087,10 @@ pub const Package = extern struct {
 
             if (from_lockfile.overrides.map.count() != to_lockfile.overrides.map.count()) {
                 summary.overrides_changed = true;
+
+                if (PackageManager.verbose_install) {
+                    Output.prettyErrorln("Overrides changed since last install", .{});
+                }
             } else {
                 for (
                     from_lockfile.overrides.map.keys(),
@@ -3096,6 +3100,9 @@ pub const Package = extern struct {
                 ) |from_k, *from_override, to_k, *to_override| {
                     if ((from_k != to_k) or (!from_override.eql(to_override, from_lockfile.buffers.string_bytes.items, to_lockfile.buffers.string_bytes.items))) {
                         summary.overrides_changed = true;
+                        if (PackageManager.verbose_install) {
+                            Output.prettyErrorln("Overrides changed since last install", .{});
+                        }
                         break;
                     }
                 }
@@ -3186,6 +3193,15 @@ pub const Package = extern struct {
                                     update_requests,
                                     null,
                                 );
+
+                                if (PackageManager.verbose_install and (diff.add + diff.remove + diff.update) > 0) {
+                                    Output.prettyErrorln("Workspace package \"{s}\" has added <green>{d}<r> dependencies, removed <red>{d}<r> dependencies, and updated <cyan>{d}<r> dependencies", .{
+                                        path,
+                                        diff.add,
+                                        diff.remove,
+                                        diff.update,
+                                    });
+                                }
 
                                 break :brk !diff.hasDiffs();
                             } else false,

